### PR TITLE
feat: add container config page

### PR DESCRIPTION
**Pull request by balanceee**
The contains the following:

- Added a new page to edit container/image config
- Removed the InstanceContainerConfig/CompleteContainerConfig/ExplicitContainerConfig
- Extracted the config field in the database. (ContainerConfig, InstanceContainerConfig).
- Reworked the SecretKeyInputField. It is no secret specific anymore.

### CODE REVIEW

File: golang/internal/grpc/grpc.go [hunk 1]
```diff
@@ -40,12 +40,10 @@ type GrpcConnectionParams struct {
 	token    string
 }
 
-type (
-	DeployFunc     func(context.Context, *dogger.DeploymentLogger, *v1.DeployImageRequest, *v1.VersionData) error
-	WatchFunc      func(context.Context, string) []*common.ContainerStateItem
-	DeleteFunc     func(context.Context, string, string) error
-	SecretListFunc func(context.Context, string, string) ([]string, error)
-)
+type DeployFunc func(context.Context, *dogger.DeploymentLogger, *v1.DeployImageRequest, *v1.VersionData) error
```
Comment (polaroi8d) [inline golang/internal/grpc/grpc.go:43]: Why?

File: protobuf/proto/agent.proto [hunk 2]
```diff
@@ -75,38 +75,159 @@ message ListSecretsRequest {
  * Deploys a single container
  *
  */
-message DeployRequest {
+message Environment { repeated string env = 1; }
+
+message InstanceConfig {
+  /*
+  containerPreName, mapped into host folder structure,
```
Comment (polaroi8d) [inline protobuf/proto/agent.proto:82]: Where is `containerPreName`?

File: protobuf/proto/agent.proto [hunk 3]
```diff
@@ -75,38 +75,159 @@ message ListSecretsRequest {
...
+  repeated VolumeLink volumes = 1000;
+  repeated string command = 1001;
+  repeated string args = 1002;
+  map<string, string> environments = 1003;
+}
+
+message ImportContainer {
+  string volume = 100;
+  string command = 101;
+
```
Comment (polaroi8d) [inline protobuf/proto/agent.proto:166]: Inconsistent newline between simple and repeated values. (It's minor but we have to fix this in the following PRs)

Comment (polaroi8d) [reply protobuf/proto/agent.proto:166]: Same with the one-line and multiple-line comments.

File: web/crux/proto/crux.proto [hunk 4]
```diff
@@ -520,13 +519,128 @@ message IncreaseVersionRequest {
...
+  optional common.ResourceConfig resourceConfig = 102;
+  optional bool proxyHeaders = 103;
+  optional bool useLoadBalancer = 104;
+
+  repeated common.UniqueKey customHeaders = 1000;
+  repeated UniqueKeyValue extraLBAnnotations = 1001;
+}
+
+message CommonContainerConfig {
+  optional string name = 101;
```
Comment (polaroi8d) [inline web/crux/proto/crux.proto:620]: `name` is optional?

File: web/crux-ui/src/components/products/versions/images/edit-image-json.tsx [hunk 5]
```diff
@@ -3,60 +3,115 @@ import useMultiInputState from '@app/components/editor/use-multi-input-state'
...
+const simplify = <T,>(item: T): Omit<T, 'id'> => {
+  const newItem: any = { ...item }
+  delete newItem.id
+
+  return newItem
+}
 
-  if (imageConfig.environment) {
-    config.environment = keyValueArrayToJson(imageConfig.environment)
+const imageConfigToJsonContainerConfig = (
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/edit-image-json.tsx:45]: We have to implement this function inside an `edit-image-json` component?

File: golang/internal/grpc/grpc.go [hunk 6]
```diff
@@ -439,7 +437,7 @@ func executeSecretList(
 		Keys:      keys,
 	}
 
-	_, err = grpcConn.Client.SecretsList(ctx, resp)
+	_, err = grpcConn.Client.GetSecretList(ctx, resp)
```
Comment (robot9706) [inline golang/internal/grpc/grpc.go:440]: Why is this renamed to `GetSecretList`? It was called `SecretList` because the agent sends data to crux, so it's not 'getting' anything.

File: web/crux-ui/src/components/products/versions/images/edit-image-json.tsx [hunk 7]
```diff
@@ -82,61 +137,168 @@ const mergeKeyValuesWithJson = (items: UniqueKeyValue[], json: Record<string, st
...
+    onPatch(merge(config, local))
 
     return local
   }
 
   const [editorState, editorActions] = useMultiInputState({
     id: EDITOR_ID,
-    value: imageConfigToCompleteContainerConfig(null, config),
+    value: instanceEditor
+      ? imageConfigToInstanceJsonConfig(null, config)
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/edit-image-json.tsx:291]: `imageConfigToInstanceJsonConfig` rename to `imageConfigToJsonInstanceConfig` like above function.

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    'importContainer',
+    'configContainer',
+    'ports',
+    'portRanges',
+    'volumes',
+    'commands',
+    'args',
+    'initContainers',
+  ],
+  kubernetes: [
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:29]: In proto we use Dagent and Crane, I just think this is a good approach to use here in a different way. A different abstraction layer, I know but it will be misunderstood in a long run, IMHO. What do you think?

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  useEffect(() => onChange(filters), [filters, onChange])
+
+  // Render
+
+  const getBorderColor = (type: BaseImageConfigFilterType): string => {
+    switch (type) {
+      case 'common':
+        return 'border-orange-400/50'
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:76]: Would be better if we predefine these colors in tailwind config like `border-dagent` and so on... No?

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  const onFilterChanged = (value: ImageConfigFilterType) => {
+    const newFilters = filters.indexOf(value) !== -1 ? filters.filter(it => it !== value) : [...filters, value]
+    setFilters(newFilters)
+  }
+
+  useEffect(() => onChange(filters), [filters, onChange])
+
+  // Render
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:71]: I don't really understand this random comment here.

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      case 'kubernetes':
+        return 'border-violet-400/50'
+      case 'docker':
+        return 'border-sky-400/50'
+      default:
+        return 'border-dyo-turquoise'
+    }
+  }
+
+  const getBgColor = (type: BaseImageConfigFilterType): string => {
```
Comment (polaroi8d) [inline web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:86]: The 2 function looks like same.

File: web/crux-ui/src/models/container.ts [hunk 12]
```diff
@@ -36,191 +38,302 @@ export type PortRange = {
...
+  'json-file',
+  'syslog',
+  'journald',
+  'gelf',
+  'fluentd',
+  'awslogs',
+  'splunk',
+  'etwlogs',
+  'logentries',
+] as const
```
Comment (polaroi8d) [inline web/crux-ui/src/models/container.ts:92]: Missing new line. Have you checked the code with the linter?

File: web/crux-ui/src/models/container.ts [hunk 13]
```diff
@@ -36,191 +38,302 @@ export type PortRange = {
...
+  tty?: boolean
+  importContainer?: ContainerConfigImportContainer
+  configContainer?: ContainerConfigContainer
+  ports?: ContainerConfigPort[]
+  portRanges?: ContainerConfigPortRange[]
+  volumes?: ContainerConfigVolume[]
+  commands?: UniqueKey[]
+  args?: UniqueKey[]
+  initContainers?: InitContainer[]
+  capabilities: UniqueKeyValue[]
```
Comment (polaroi8d) [inline web/crux-ui/src/models/container.ts:164]: Are `capabilities` the only value that is mandatory?

File: web/crux-ui/src/models/container.ts (hunk 12 above)
Comment (balanceee) [reply web/crux-ui/src/models/container.ts:92]: I think its a github bug. There is a new line.

File: web/crux/proto/crux.proto (hunk 4 above)
Comment (balanceee) [reply web/crux/proto/crux.proto:620]: Not supposed to be. TY

File: web/crux-ui/src/components/products/versions/images/config/crane-config-section.tsx [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+  const { config: propsConfig, filters, onChange: propsOnChange, editorOptions } = props
+
+  const [config, setConfig] = useState<CraneConfigDetails>(propsConfig)
+
+  const onChange = (newConfig: Partial<ContainerConfig>) => {
+    setConfig({ ...config, ...newConfig })
+    propsOnChange(newConfig)
+  }
+
+  const toNumber = (value: string): number => {
```
Comment (robot9706) [inline web/crux-ui/src/components/products/versions/images/config/crane-config-section.tsx:38]: This is also defined in CommonConfigSection, so I think it'd be better to move it to a utils file.

File: web/crux-ui/src/components/products/versions/images/config/common-config-section.tsx [hunk 15]
```diff
@@ -0,0 +1,801 @@
...
+
+    return Number.isNaN(value) ? 0 : Number(value)
+  }
+
+  const onChange = (newConfig: Partial<ContainerConfig>) => {
+    setConfig({ ...config, ...newConfig })
+    propsOnChange(newConfig)
+  }
+
+  const onComplexValueChanged = <T,>(property: string, field: string, value: T, index: number) => {
```
Comment (robot9706) [inline web/crux-ui/src/components/products/versions/images/config/common-config-section.tsx:60]: `onChange` few lines above, so `onComplexValueChange` would be more fitting :)

File: web/crux-ui/src/components/products/versions/images/edit-image-json.tsx [hunk 16]
```diff
@@ -3,60 +3,115 @@ import useMultiInputState from '@app/components/editor/use-multi-input-state'
...
-    if (!prev[it.key]) {
-      prev[it.key] = it.value
-    }
+type JsonConfig = InstanceJsonContainerConfig | JsonContainerConfig
+
+const keyValueArrayToJson = (list: UniqueKeyValue[]): JsonKeyValue => {
+  const result = {}
 
-    return prev
+  list?.forEach(it => {
```
Comment (robot9706) [inline web/crux-ui/src/components/products/versions/images/edit-image-json.tsx:29]: This could be done with a one-liner, something like:

```
return list.reduce((prev, it) => ({ ...prev, [it.key]: it.value }), {})
```

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx (hunk 10 above)
Comment (robot9706) [inline web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:71]: Is this comment necessary?

File: web/crux-ui/src/validations/deployment.ts [hunk 17]
```diff
@@ -17,12 +15,12 @@ export const createDeploymentSchema = updateDeploymentSchema.concat(
...
-export const deploymentSchema = yup.object().shape({
-  environment: uniqueKeyValuesSchema,
-  instances: yup.array(
-    yup.object().shape({
-      image: imageSchema,
-      overriddenConfig: explicitContainerConfigSchema.nullable(),
-    }),
-  ),
-})
+// export const deploymentSchema = yup.object().shape({
```
Comment (robot9706) [inline web/crux-ui/src/validations/deployment.ts:18]: Comment? 🌊

**Review by robot9706 — CHANGES_REQUESTED** [review]

**Review by nandor-magyar — CHANGES_REQUESTED** [review]
I've checked go part thoroughly and skimmed the others. See my comments. The more the change, the more respect it is. :gun:

File: web/crux-ui/src/components/products/versions/images/image-config-filters.tsx (hunk 11 above)
Comment (polaroi8d) [reply web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:86]: Okay the name is not the same but doing the same inside.

Comment (balanceee) [reply web/crux-ui/src/components/products/versions/images/image-config-filters.tsx:86]: Because of the tailwind's tree shaking we have to write down the whole class name so it cant be simplified

File: golang/api/v1/deploy.go [hunk 18]
```diff
@@ -173,6 +173,8 @@ type ContainerConfig struct {
 	RestartPolicy builder.RestartPolicyName `json:"restartPolicy"`
 	// bridge(container, default) host, none or network name
 	NetworkMode string `json:"networkMode"`
+	// extra networks
```
[the remark this answers is not in the document]
Comment (balanceee) [reply golang/api/v1/deploy.go:176]: ?

File: golang/internal/grpc/grpc.go (hunk 6 above)
Comment (balanceee) [reply golang/internal/grpc/grpc.go:440]: Bad merge

File: web/crux-ui/src/components/products/versions/images/edit-image-json.tsx (hunk 5 above)
Comment (balanceee) [reply web/crux-ui/src/components/products/versions/images/edit-image-json.tsx:45]: Only used in this file.

File: web/crux/src/app/deploy/deploy.mapper.ts [hunk 19]
```diff
@@ -212,7 +193,99 @@ export default class DeployMapper {
     return state ? (containerStateToJSON(state).toLowerCase() as ContainerStateEnum) : null
   }
 
+  configToCommonConfig(config: ContainerConfigData): CommonContainerConfig {
+    return {
+      name: config.name,
+      environments: this.jsonToPipedFormat(config.environment as JsonArray),
```
[the remark this answers is not in the document]
Comment (m8vago) [reply web/crux/src/app/deploy/deploy.mapper.ts:199]: @nandor-magyar

**Review by m8vago — CHANGES_REQUESTED** [review]

File: golang/api/v1/deploy.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (peterambrus3) [reply golang/api/v1/deploy.go:176]: it was dagent-only but someone moved it to k8s-only section

File: golang/internal/dogger/dogger_test.go [hunk 20]
```diff
@@ -26,7 +25,7 @@ func (tStream *AgentTestStream) Send(message *common.DeploymentStatusMessage) er
 	return nil
 }
 
-func (tStream *AgentTestStream) CloseAndRecv() (*agent.Empty, error) {
+func (tStream *AgentTestStream) CloseAndRecv() (*common.Empty, error) {
 	return nil, nil
```
[the remark this answers is not in the document]
Comment (peterambrus3) [reply golang/internal/dogger/dogger_test.go:29]: thats not our code we just renamed the refactored config namespace, maybe @nandor-magyar can help here ?

File: web/crux-ui/locales/en/container.json [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    "path": "Path",
+    "class": "Class (k8s)",
+    "type": "Type",
+    "volumeTypes": { "ro": "Ro", "rw": "Rw", "rwx": "Rwx", "mem": "Mem", "tmp": "Tmp" },
+    "files": "Files",
+    "commands": "Commands",
+    "arguments": "Arguments",
+    "ttySupport": "Tty support",
+    "exposeStrategy": "Expose strategy",
+    "exposeStrategies": { "none": "None", "expose": "Expose", "expose_with_tls": "Expose with tls" },
```
[the remark this answers is not in the document]
Comment (peterambrus3) [reply web/crux-ui/locales/en/container.json:27]: probably my bad, we will fix it thx

File: web/crux-ui/src/components/shared/key-value-input.tsx [hunk 22]
```diff
@@ -153,15 +173,13 @@ const KeyValueInput = (props: KeyValueInputProps) => {
   }
 
   return (
-    <form className={clsx(className, 'flex flex-col max-h-128 overflow-y-auto')}>
-      {!heading ? null : (
-        <DyoHeading element="h6" className="text-bright mt-4 mb-2">
-          {heading}
-        </DyoHeading>
+    <div className={clsx(className, 'flex flex-col max-h-128 overflow-y-auto')}>
```
[the remark this answers is not in the document]
Comment (peterambrus3) [reply web/crux-ui/src/components/shared/key-value-input.tsx:176]: good point, we should discuss this in person

File: web/crux-ui/src/server/crux/mappers/image-mappers.ts [hunk 23]
```diff
@@ -1,124 +1,212 @@
...
 
-  return { driver: logConfig.type, options: logConfig.config }
+  return {
+    driver: logConfig.driver ? driverTypeFromJSON(logConfig.driver.toUpperCase()) : DriverType.DRIVER_TYPE_NONE,
+    options: logConfig.options,
+  }
+}
+
+export const logDriverDto = (logDriver?: DriverType): ContainerLogDriverType => {
+  if (logDriver == null) {
```
[the remark this answers is not in the document]
Comment (peterambrus3) [reply web/crux-ui/src/server/crux/mappers/image-mappers.ts:69]: true, also have to change the logConfigToDto function (null case) below

File: golang/api/v1/deploy.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (nandor-magyar) [reply golang/api/v1/deploy.go:176]: Ah, it is the correct location like this, my bad.

File: web/crux/src/app/deploy/deploy.mapper.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (nandor-magyar) [reply web/crux/src/app/deploy/deploy.mapper.ts:199]: Because we still have the "core" in use.

File: golang/internal/dogger/dogger_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (nandor-magyar) [reply golang/internal/dogger/dogger_test.go:29]: I don't remember anything specific, if not in use by use ( as anywhere in the repo) it could be removed.

**Comment by polaroi8d** [comment]
Missing E2E tests 🤡

**Review by polaroi8d — APPROVED** [review]
Amazing job @balanceee and @peterambrus3 🔥

File: web/crux-ui/src/utils.ts [hunk 24]
```diff
@@ -365,3 +365,28 @@ export const toTimestamp = (date: Date): Timestamp => {
...
+      break
+    }
+  }
+
+  return !empty ? object : null
+}
+
+export const toNumber = (value: string): number => {
+  if (!value) {
+    return undefined
```
Comment (m8vago) [inline web/crux-ui/src/utils.ts:388]: please, return null instead

**Review by m8vago — APPROVED** [review]
